### DISCUSSION TITLE
Collecting metrics by time

### DISCUSSION THREAD

**Discussion by kvartik** [yabeda-rb/yabeda#28] [Q&A]
Hi
please help with the block to collect metrics by time 

```Yabeda.configure do
  # This block will be executed periodically few times in a minute
  # (by timer or external request depending on adapter you're using)
  # Keep it fast and simple!
  collect do
    your_app.whistles_active.set({}, Whistle.where(state: :active).count)
  end
end
```

how to specify the time of metrics assembly in this block?
so that the block is not executed at every request

**Comment by Envek** [comment]
Frequency of `collect` blocks invocation is adapter-specific:
 - `yabeda-prometheus` executes them on every scraping request (see [its source](https://github.com/yabeda-rb/yabeda-prometheus/blob/4e9e4ca60cb90ef3f8ec95671dfd82957fd625fb/lib/yabeda/prometheus/exporter.rb#L48)). It can't be configured, but you can make scraping less frequent in Prometheus scraping config: default of 10 seconds can be too often).
 - `yabeda-newrelic` executes them on every harvesting cycle (see [its source](https://github.com/yabeda-rb/yabeda-newrelic/blob/6ef8995816257393a4004ac7b7050483fe5eadb5/lib/yabeda/newrelic/configure.rb#L7-L9)). Can't be configured.
 - `yabeda-datadog` makes it manually via auxiliary thread (see [its source](https://github.com/yabeda-rb/yabeda-datadog/blob/75304b67163630dde198ebddc785510f5b889884/lib/yabeda/datadog.rb#L45-L50)). Can't be configured.

---

I suppose that you're using yabeda-prometheus.

If you want to change its behavior, you can subclass [its exporter class](https://github.com/yabeda-rb/yabeda-prometheus/blob/4e9e4ca60cb90ef3f8ec95671dfd82957fd625fb/lib/yabeda/prometheus/exporter.rb), re-implement its initialization and `call` method, and start a background thread on init to collect metrics periodically instead of calling collection on every `call` (see [example of such implementation](https://github.com/yabeda-rb/yabeda-prometheus2influxdb/blob/master/lib/yabeda/prometheus2influxdb.rb#L34-L42)). Then just use your exporter subclass in place of ours.

---

BTW, can you please tell a bit more about your use case? Why you want control that?
